core/src/main/scala/cats/data/Kleisli.scala

```diff
@@ -93,19 +98,22 @@ final case class Kleisli[F[_], -A, B](run: A => F[B]) { self =>
    * }}}
    */
   def andThen[C](k: Kleisli[F, B, C])(implicit F: FlatMap[F]): Kleisli[F, A, C] =
-    this.andThen(k.run)
+    k.run match {
+      case _: StrictConstFunction1[_] => k.asInstanceOf[Kleisli[F, A, C]]
+      case _                          => this.andThen(k.run)
+    }
```

Comment:
TODO: revert this change. Will prevent effect in `this` from being evaluated at all.